File: 1261-find-elements-in-a-contaminated-binary-tree/1261-find-elements-in-a-contaminated-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
// var FindElements = function(root) {
//     this.root = this.recover(root, 0)
// };

// FindElements.prototype.recover = function(root, x) {
//     if (!root) return null
//     root.val = x
//     root.left = this.recover(root.left, 2 * x + 1)
//     root.right = this.recover(root.right, 2 * x + 2)
//     return root
// }

/** 
 * @param {number} target
 * @return {boolean}
 */
// FindElements.prototype.find = function(target) {
//     return dfs(this.root, target)
    
//     function dfs(root, target) {
//         if (!root) return false
//         if (root.val === target) return true
//         return dfs(root.left, target) || dfs(root.right, target)
//     }
// };

/** 
 * Your FindElements object will be instantiated and called as such:
 * var obj = new FindElements(root)
 * var param_1 = obj.find(target)
 */

class FindElements {
    constructor(root) {
        this.set = new Set()
        this.root = this.recover(root, 0)
    }
    // recover(root, x) {
    //     if (!root) return null
    //     root.val = x
    //     this.set.add(x)
    //     root.left = this.recover(root.left, 2 * x + 1)
    //     root.right = this.recover(root.right, 2 * x + 2)
    //     return root
    // }
    recover(root, x) {
        const q = [[root, x]]
        while (q.length) {
            const [root, x] = q.shift()
            root.val = x
            this.set.add(x)
            if (root.left) q.push([root.left, 2 * x + 1])
            if (root.right) q.push([root.right, 2 * x + 2])
        }
        return root
    }
    find(target) {
        return this.set.has(target)
    }
}

File: 1261-find-elements-in-a-contaminated-binary-tree/1261-find-elements-in-a-contaminated-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class FindElements:

    def __init__(self, root: Optional[TreeNode]):
        self.set = set()
        self.root = self.recover(root, 0)

    def find(self, target: int) -> bool:
        return target in self.set
        
    def recover(self, root: Optional[TreeNode], x: int) -> TreeNode:
        if not root:
            return None
        root.val = x
        self.set.add(x)
        root.left = self.recover(root.left, 2 * x + 1)
        root.right = self.recover(root.right, 2 * x + 2)
        return root
        


# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)